pkg/provider/chef/chef.go

```diff
@@ -0,0 +1,314 @@
+	errInvalidURL                            = "invalid serverurl: %w"
+	errChefClient                            = "unable to create chef client: %w"
+	errChefProvider                          = "missing or invalid spec: %w"
+	errUninitalizedChefProvider              = "chef provider is not initialized"
+	errNoDatabagItemFound                    = "data bag item %s not found in data bag %s"
+	errNoDatabagItemPropertyFound            = "property %s not found in data bag item"
+	errCannotListDataBagItems                = "unable to list items in data bag %s"
+	errUnableToConvertToJSON                 = "unable to convert databagItem into JSON"
+	errInvalidFormat                         = "invalid key format in data section. Expected value 'databagName/databagItemName'"
+	errStoreValidateFailed                   = "unable to validate provided store. Check if username, serverUrl and privateKey are correct"
+	errServerURLNoEndSlash                   = "serverurl does not end with slash(/)"
+	errInvalidDataform                       = "invalid key format in dataForm section. Expected only 'databagName'"
+
+	ProviderChef             = "Chef"
+	CallChefGetDataBagItem   = "GetDataBagItem"
+	CallChefListDataBagItems = "ListDataBagItems"
+	CallChefGetUser          = "GetUser"
+)
+
+type DatabagFetcher interface {
+	GetItem(databagName string, databagItem string) (item chef.DataBagItem, err error)
+	ListItems(name string) (data *chef.DataBagListResult, err error)
+}
+
+type UserInterface interface {
+	Get(name string) (user chef.User, err error)
+}
+
+type Providerchef struct {
+	clientName     string
+	databagService DatabagFetcher
+	userService    UserInterface
+}
+
+var _ v1beta1.SecretsClient = &Providerchef{}
+var _ v1beta1.Provider = &Providerchef{}
+
+var log = ctrl.Log.WithName("provider").WithName("chef").WithName("secretsmanager")
+
+func init() {
+	v1beta1.Register(&Providerchef{}, &v1beta1.SecretStoreProvider{
+		Chef: &v1beta1.ChefProvider{},
+	})
+}
+
+func (providerchef *Providerchef) NewClient(ctx context.Context, store v1beta1.GenericStore, kube kclient.Client, namespace string) (v1beta1.SecretsClient, error) {
+	chefProvider, err := getChefProvider(store)
+	if err != nil {
+		return nil, fmt.Errorf(errChefProvider, err)
+	}
+
+	credentialsSecret := &corev1.Secret{}
+	objectKey := types.NamespacedName{
+		Name:      chefProvider.Auth.SecretRef.SecretKey.Name,
+		Namespace: namespace,
+	}
+
+	if store.GetObjectKind().GroupVersionKind().Kind == v1beta1.ClusterSecretStoreKind {
+		if chefProvider.Auth.SecretRef.SecretKey.Namespace == nil {
+			return nil, fmt.Errorf(errInvalidClusterStoreMissingPKNamespace)
+		}
+		objectKey.Namespace = *chefProvider.Auth.SecretRef.SecretKey.Namespace
+	}
+
+	err = kube.Get(ctx, objectKey, credentialsSecret)
+	if err != nil {
+		return nil, fmt.Errorf(errFetchK8sSecret, err)
+	}
+
+	secretKey := credentialsSecret.Data[chefProvider.Auth.SecretRef.SecretKey.Key]
+	if (secretKey == nil) || (len(secretKey) == 0) {
```

Comment:
```suggestion
	if len(secretKey) == 0 {
```

len already does nil check